[Instrumentation.AWS] Update the instrumentation logic to use AWS TracerProvider.

muhammad-othman opened the pull request:
## Changes

This PR focuses on leveraging the newly introduced tracing APIs within the AWS SDK. The main changes include:

* Integration of New Tracing APIs:
Implemented the new AWS SDK tracing APIs and updated the `AddAWSInstrumentation` method to register this implementation using `AWSTracerProvider`.

* Simplification of Tracing Logic:
Removed `AWSTracingPipelineHandler` and migrated its logic to either the .NET SDK or `AWSPropagatorPipelineHandler`.

* SDK Version Update:
Updated the SDK version to `3.7.400` to incorporate the new tracing APIs.

## Merge requirement checklist

* [x] Unit tests added/updated
* [x] Appropriate `CHANGELOG.md` files updated for non-trivial changes

Kielek commented:
@muhammad-othman, please sign EasyCLA. Without this we are not allowed to proceed with the PR.

normj commented:
With the previous implementation we had the following AWS attributes added on the span 

![image](https://github.com/user-attachments/assets/c03ec7bc-bf91-4e2c-b0ca-4bedbfd2b4c9)

And with the new implementation we are only getting queue url and request id.

![image](https://github.com/user-attachments/assets/cb488604-1e97-48ee-84c3-2a1195d17993)

Can we get operation, region and service added on the spans in the new attribute?

muhammad-othman commented:
> With the previous implementation we had the following AWS attributes added on the span
> 
> ![image](https://private-user-images.githubusercontent.com/1653751/353923792-c03ec7bc-bf91-4e2c-b0ca-4bedbfd2b4c9.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MjI1MjUzMTMsIm5iZiI6MTcyMjUyNTAxMywicGF0aCI6Ii8xNjUzNzUxLzM1MzkyMzc5Mi1jMDNlYzdiYy1iZjkxLTRlMmMtYjBjYS00YmVkYmZkMmI0YzkucG5nP1gtQW16LUFsZ29yaXRobT1BV1M0LUhNQUMtU0hBMjU2JlgtQW16LUNyZWRlbnRpYWw9QUtJQVZDT0RZTFNBNTNQUUs0WkElMkYyMDI0MDgwMSUyRnVzLWVhc3QtMSUyRnMzJTJGYXdzNF9yZXF1ZXN0JlgtQW16LURhdGU9MjAyNDA4MDFUMTUxMDEzWiZYLUFtei1FeHBpcmVzPTMwMCZYLUFtei1TaWduYXR1cmU9NWYzNjI2MTA4NzgyMDY2NmJhYTk1ZjllMjQxNmQyZjMwNGE2YTM1MzU3OTE1OTcxMDRmNzE2NDBiMGZkZDE4ZCZYLUFtei1TaWduZWRIZWFkZXJzPWhvc3QmYWN0b3JfaWQ9MCZrZXlfaWQ9MCZyZXBvX2lkPTAifQ.qtGLA7Sao4i5DTUTC0ri0sNlftOqTx9Ae8H12sSKRNI)
> 
> And with the new implementation we are only getting queue url and request id.
> 
> ![image](https://private-user-images.githubusercontent.com/1653751/353924354-8cb147c2-67e5-4954-9b48-125361a726e9.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MjI1MjUzMTMsIm5iZiI6MTcyMjUyNTAxMywicGF0aCI6Ii8xNjUzNzUxLzM1MzkyNDM1NC04Y2IxNDdjMi02N2U1LTQ5NTQtOWI0OC0xMjUzNjFhNzI2ZTkucG5nP1gtQW16LUFsZ29yaXRobT1BV1M0LUhNQUMtU0hBMjU2JlgtQW16LUNyZWRlbnRpYWw9QUtJQVZDT0RZTFNBNTNQUUs0WkElMkYyMDI0MDgwMSUyRnVzLWVhc3QtMSUyRnMzJTJGYXdzNF9yZXF1ZXN0JlgtQW16LURhdGU9MjAyNDA4MDFUMTUxMDEzWiZYLUFtei1FeHBpcmVzPTMwMCZYLUFtei1TaWduYXR1cmU9MzVjZTE1NjI0NzY2OTcwZWNjMTAxMDNkZDZlMTc1ZDY0Y2YxYmJjMmQ5MDEyMWRiYTQ3ODliNTNjNGYzYWUwZCZYLUFtei1TaWduZWRIZWFkZXJzPWhvc3QmYWN0b3JfaWQ9MCZrZXlfaWQ9MCZyZXBvX2lkPTAifQ.rN8TvGPsEu259wv9CJpxNYNWU_gFVXnmT56SSCo1I5s)
> 
> Can we get operation, region and service added on the spans in the new attribute?

@normj 
Yep, the following attributes was removed since they aren't part of OTel specs

![image](https://github.com/user-attachments/assets/d486b1e9-84bd-4fc7-a178-8a2b52d1bfab)


I believe they were added them preemptively before they were added to the specs, however now we have  [rpc.method](https://opentelemetry.io/docs/specs/semconv/attributes-registry/rpc/) and [rpc.service](https://opentelemetry.io/docs/specs/semconv/attributes-registry/rpc/) to replace `aws.service` and `aws.operation` attributes 
https://opentelemetry.io/docs/specs/semconv/cloud-providers/aws-sdk/#common-attributes

The only attribute that I removed is `aws.region` which doesn't have any replacement in the specs yet, not sure if we should keep it knowing that it may be added in a different name to specs, and if we will keep it, shouldn't it get added to the other SDKs too?

normj commented:
> I believe they were added them preemptively before they were added to the specs, however now we have [rpc.method](https://opentelemetry.io/docs/specs/semconv/attributes-registry/rpc/) and [rpc.service](https://opentelemetry.io/docs/specs/semconv/attributes-registry/rpc/) to replace aws.service and aws.operation attributes
https://opentelemetry.io/docs/specs/semconv/cloud-providers/aws-sdk/#common-attributes

I missed that the service and operation were moved to OTel specific attributes. Region probably doesn't make sense at the OTel spec level since that is more of an AWS attribute. We should discuss with the rest of the SDK if we should add an `aws.region` attribute to our standardization. That isn't really tied to this PR so we don't need to hold this PR back for that.

normj commented on test/OpenTelemetry.Instrumentation.AWS.Tests/TestAWSClientInstrumentation.cs:
```diff
@@ -207,9 +207,6 @@ private void ValidateAWSActivity(Activity aws_activity, Activity parent)
     private void ValidateDynamoActivityTags(Activity ddb_activity)
     {
         Assert.Equal("DynamoDB.Scan", ddb_activity.DisplayName);
-        Assert.Equal("DynamoDB", Utils.GetTagValue(ddb_activity, "aws.service"));
```
Instead of removing these checks for `aws.service` and `aws.operation` shouldn't they updated to check `rpc.service` and `rpc.method`?

muhammad-othman replied:
Yep, they do exist few lines below
![image](https://github.com/user-attachments/assets/a5d31edd-848c-4c51-bd5f-3b7143c9ec53)

normj commented on test/OpenTelemetry.Instrumentation.AWS.Tests/TestAWSClientInstrumentation.cs:
```diff
@@ -220,9 +217,6 @@ private void ValidateDynamoActivityTags(Activity ddb_activity)
     private void ValidateSqsActivityTags(Activity sqs_activity)
     {
         Assert.Equal("SQS.SendMessage", sqs_activity.DisplayName);
-        Assert.Equal("SQS", Utils.GetTagValue(sqs_activity, "aws.service"));
```
Instead of removing these checks for `aws.service` and `aws.operation` shouldn't they updated to check `rpc.service` and `rpc.method`?

muhammad-othman replied:
Yep, they do exist few lines below
![image](https://github.com/user-attachments/assets/a5d31edd-848c-4c51-bd5f-3b7143c9ec53)

cijothomas commented:
@srprash Could you review?

martincostello commented on src/OpenTelemetry.Instrumentation.AWS/Implementation/Tracing/AWSTraceSpan.cs:
```diff
@@ -0,0 +1,91 @@
+
+namespace OpenTelemetry.Instrumentation.AWS.Implementation.Tracing;
+
+internal class AWSTraceSpan : TraceSpan
```
Make `AWSTraceSpan`, `AWSTracer` and `AWSTracerProvider` `sealed` to aid with JIT optimisations?

birojnayak commented:
`aws.region` is missing as pointed out earlier, while others have replacement.  See if you can have, because existing customers/users might be relying on it to build custom dashboard or extensions.

muhammad-othman commented:
@birojnayak Unfortunately  `aws.region` attribute doesn't currently exist in the other AWS SDKs and there is no replacement for it in the current specifications. My concern is that if we add this attribute now, we may later reach an agreement with other SDKs on a different name, similar to how `aws.service` and `aws.operation` were replaced by `rpc.service and `rpc.method`. 

Since that we're still in beta, I believe it's better to remove `aws.region` for now and we can add it later with a name that has been agreed upon across all SDKs. 
Also this change will not be part of this PR since this package isn't producing the shared attributes anymore, it will be added to the Core SDK similar to https://github.com/aws/aws-sdk-net/blob/2cd501b0108ea1c0136094c04c3e1785ff0f0aac/sdk/src/Core/Amazon.Runtime/Telemetry/Tracing/TracingUtilities.cs#L52

muhammad-othman commented:
Hi @cijothomas,
I've addressed the comments on this PR and the two component owners approved it, please let me know if there are any additional steps needed from my side to merge it.

cijothomas commented:
There is a CI failure: https://github.com/open-telemetry/opentelemetry-dotnet-contrib/actions/runs/10498004987/job/29082060927?pr=1974 Could you check and fix? I can merge once CI is passing.

muhammad-othman commented:
Thank you @cijothomas for the quick response.
Looks like exported items were mutated when other tests run at the same time, which caused `Collection was modified` exception.
I moved the exportItems asserts outside the using clause, this way we are sure the tracer provider was disposed before enumerating over the exportedItems and they shouldn't be modified at this stage.

cijothomas commented:
@vishweshbankwar Need your help. One of the merge requirement is not met, maybe a CI is missing? I am not able to tell quickly..

muhammad-othman commented:
@cijothomas I think it needs code owner approver first, someone from ` open-telemetry/dotnet-contrib-approvers`. I believe [the previous PR](https://github.com/open-telemetry/opentelemetry-dotnet-contrib/pull/1980) was approved by you then `Code owner review required` disappeared.
